app/services/oidc_jwt_token_validate.rb

```diff
@@ -0,0 +1,54 @@
+class OidcJwtTokenValidate
+  attr_reader :decoded_token
+  delegate :logger, to: :Rails
+
+  def initialize(jwt_token)
+    @jwt_token = jwt_token
+  end
+
+  def valid_jwt?
+    valid_signature? && type_bearer? && valid_aud?
+  rescue JWT::InvalidAudError
+    logger.error "OidcJwtToken SSO: Invalid token."
+    false
+  rescue JWT::ImmatureSignature
+    logger.error "OidcJwtToken SSO: JWT is for the future."
+    false
+  rescue JWT::ExpiredSignature
+    logger.error "OidcJwtToken SSO: Signature has expired."
+    false
+  rescue JWT::DecodeError
+    logger.error "OidcJwtToken SSO: Failed to decode JWT."
+    false
+  end
+
+  private
+
+  def generate_jwks_uri
+    response = RestClient::Request.execute(
+      :url => Setting['keycloak_jwks_url'],
+      :method => :get,
+      :headers => { content_type: 'application/x-www-form-urlencoded' },
+      :verify_ssl => true
+    )
+    rsa_public = JSON.parse(response)['keys'].first
+    jwk = JSON::JWK.new(rsa_public)
+    jwk.to_key
+  rescue RestClient::Exception, SocketError, JSON::ParserError => e
+    logger.error e
+    nil
+  end
+
+  def valid_signature?
+    @decoded_token = JWT.decode(@jwt_token, generate_jwks_uri, true, { algorithm: 'RS256' }).first
```

Comment:
What happens if `generate_jwks_uri` returns `nil`? Can we please have a test for this? (What I actually want to have is a test when there is an error retrieving the `keycloak_jwks_url` data)

Reply:
I think this one should do the trick: https://github.com/theforeman/foreman/pull/6549/files#diff-6e97738229d9729318212fd628b7cd9aR49

Let me know if this is what you were expecting.

Reply:
@timogoebel  added a few more [here](https://github.com/theforeman/foreman/pull/6549/files#diff-6e97738229d9729318212fd628b7cd9aR57), hope that makes it clear :)